jonason/Parser.cpp

```diff
@@ -0,0 +1,226 @@
+//
+//  Parser.cpp
+//  jonason
+//
+//  Created by Gabriel Lumbi on 2023-03-03.
+//
+
+#include "Parser.hpp"
+
+#include "Tokenizer.hpp"
+
+namespace jonason {
+
+// Errors
+
+const ParseError ParseError::unexpected_eof(ParseError::UNEXPECTED_EOF, std::string("Wow there, I didn't expect this end-of-file!"));
+
+ParseError ParseError::unexpected_token(const Token& token) {
+    if (token.tag == Token::LITERAL) {
+        return ParseError(ParseError::UNEXPECTED_TOKEN, std::string("Did quite get what you meant by this buddy: ") + token.value);
+    } else {
+        return ParseError(ParseError::UNEXPECTED_TOKEN, std::string("Hey, is this a typo?: ") + static_cast<char>(token.tag));
+    }
+}
+
+// Parsers
+
+using Iterator = std::vector<Token>::iterator&;
+using Out = JSONValue*;
+
+static void parse_json_token(Token::Type, Iterator);
+static Out parse_json_value(Iterator, Iterator);
+static Out parse_json_object(Iterator, Iterator);
+static std::pair<std::string, JSONValue*> parse_json_object_key_value(Iterator, Iterator);
+static JSONValue::KeyType parse_json_object_key(Iterator, Iterator);
+static Out parse_json_array(Iterator, Iterator);
+static Out parse_json_string(Iterator, Iterator);
+
+std::unique_ptr<JSONValue> parse(const std::string& string)
+{
+    if (string.empty()) { throw ParseError::unexpected_eof; }
+
+    std::vector<Token> tokens;
+    tokenize(string, tokens);
+
+    if (tokens.empty()) { throw ParseError::unexpected_eof; }
+
+    auto begin = tokens.begin();
+    auto end = tokens.end();
+    return std::unique_ptr<JSONValue>(parse_json_value(begin, end));
+}
+
+std::unique_ptr<JSONValue> parse(std::istream& istream)
+{
+    std::vector<Token> tokens;
+    tokenize(istream, tokens);
+
+    if (tokens.empty()) { throw ParseError::unexpected_eof; }
+
+    auto begin = tokens.begin();
+    auto end = tokens.end();
+    return std::unique_ptr<JSONValue>(parse_json_value(begin, end));
+}
+
+static void parse_json_token(Token::Type token, Iterator iterator)
+{
+    if (iterator->tag == token) {
+        iterator++;
+    } else {
+        throw ParseError::unexpected_token(*iterator);
+    }
+}
+
+Out parse_json_value(Iterator iterator, Iterator end)
+{
+    switch (iterator->tag) {
+        case Token::DOUBLE_QUOTE:
+            return parse_json_string(iterator, end);
+        case Token::LITERAL:
+            switch (*iterator->value) {
+                case 't':
+                    if (strcmp(iterator->value, "true") == 0) {
+                        iterator++;
+                        return new JSONValue(true);
```

Comment:
Returning a raw pointer feels evil but I was worried about the overhead of using smart pointers. What would be a better way?

Reply:
There is no overhead to using smart pointers, it wraps the same operations you'd do on the pointer. Maybe you'll get an additional `if (!nullptr)` in the destructor to handle the moved case, but even that can probably be optimized away.